File: 109_convert_sorted_list_to_binary_search_tree/sortedListToBST.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        root = self.dfs(head)
        return root

    def dfs(self, start):
        if not start: return None        
        #if start == end: return TreeNode(start.val)

        prev_slow = None # the node that one node before the slow
        slow = start # move one nodes each time
        fast = start # move two nodes each time

        # slow node will stop at middle (root node)
        while fast.next and fast.next.next:
            prev_slow = slow
            fast = fast.next.next
            slow = slow.next
        
        # slow is the root at this level
        tree_node = TreeNode(slow.val)

        # using slow as pivot, dividing the linked list
        # into two sub-list (:pre_slow), (slow.next:)
        if prev_slow:
            prev_slow.next = None
            tree_node.left = self.dfs(start)
        
        if slow.next:
            tree_node.right = self.dfs(slow.next)

        return tree_node

File: 109_convert_sorted_list_to_binary_search_tree/sortedListToBST.cpp
#include <iostream>

using namespace std;
//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


//Definition for a binary tree node.
struct TreeNode {
    int val;
	TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
    public:
        TreeNode* sortedListToBST(ListNode* head) {
            return dfs(head);
            
        }
    private:
        TreeNode* dfs(ListNode* start) {
            if(!start) return NULL;
            ListNode* prevSlow = NULL;
            ListNode* slow = start;
            ListNode* fast = start;

            while(fast->next && fast->next->next) {
                prevSlow = slow;
                slow = slow -> next;
                fast = fast -> next -> next;
            }

            TreeNode* treeNode = new TreeNode(slow->val);

            if(prevSlow) {
                prevSlow -> next = NULL;
                treeNode -> left = dfs(start);
            }
            if(slow->next) {
                treeNode -> right = dfs(slow -> next);
            }
            return treeNode;
        }

};


int main() {
    return 0;
}
